### PR TITLE
test(cargo): verify unmap operation for aio module

### PR DESCRIPTION
**Pull request by cjones1024**
Test is currently disabled until spdk is updated.

### CODE REVIEW

File: mayastor/tests/aio_unmap.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+                .unwrap();
+        }
+
+        let pool = Lvs::lookup("tpool").unwrap();
+        assert_eq!(pool.lvols().unwrap().count(), 4);
+    })
+    .await;
+
+    // verify that some blocks have been allocated
+    assert_ne!(get_allocated_blocks(DISKNAME1).unwrap(), 0);
```
Comment (tiagolobocastro) [inline mayastor/tests/aio_unmap.rs:76]: Looks like these lvols only have metadata, should we also write some "user data" to the lvol?
Would it be worth sending explicit unmap to each lvol or even via a nexus? (to cover the case for unmaps from the pod host)?

File: mayastor/tests/aio_unmap.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    .await;
+
+    // Check that we're able to find our new LVS.
+    ms.spawn(async {
+        assert_eq!(Lvs::iter().count(), 1);
+        let pool = Lvs::lookup("tpool").unwrap();
+        assert_eq!(pool.name(), "tpool");
+        assert_eq!(pool.used(), 0);
+        dbg!(pool.uuid());
+        assert_eq!(pool.base_bdev().name(), "/tmp/disk1.img");
```
Comment (jonathan-teh) [inline mayastor/tests/aio_unmap.rs:57]: Use DISKNAME here.

File: mayastor/tests/aio_unmap.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        assert_eq!(Lvs::iter().count(), 0);
+
+        // no bdevs
+        assert_eq!(Bdev::bdev_first().into_iter().count(), 0);
+    })
+    .await;
+
+    // Verify that all used blocks have been discarded by confirming
+    // that no blocks are currently allocated for the backing store.
+    assert_eq!(get_allocated_blocks(DISKNAME1).unwrap(), 0);
```
Comment (jonathan-teh) [inline mayastor/tests/aio_unmap.rs:115]: Use a custom message with the assert so it's clear if it fails and the assert itself becomes self-documenting.

File: mayastor/tests/aio_unmap.rs (hunk 1 above)
Comment (cjones1024) [reply mayastor/tests/aio_unmap.rs:76]: Yes we could write user data to the lvols. However this was not supposed to be a test to validate that the upper layers do the right thing when the lvol or pool is destroyed. This was intended to be a simple test to validate that when the upper layers want to zero a region, the blocks comprising that region are in fact unallocated. The fact that everything is zeroed means I don't need to complicate things by having to verify precisely where a hole is located in a file.

Comment (cjones1024) [reply mayastor/tests/aio_unmap.rs:76]: After discussing with tiago, will create 2 tests:
- low level tests that verifies calling unmap on a bdev actually unallocates the blocks form the underlying file.
- a higher level test that verifies that destroying an lvol or pool does in fact zero the data (without being concerned as to how the zeroes are written).

File: mayastor/tests/aio_unmap.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        assert_eq!(pool.used(), 0);
+        dbg!(pool.uuid());
+        assert_eq!(pool.base_bdev().name(), "/tmp/disk1.img");
+    })
+    .await;
+
+    // Create 4 lvols on this pool.
+    ms.spawn(async {
+        let pool = Lvs::lookup("tpool").unwrap();
+        for i in 0 .. 4 {
```
Comment (mtzaurus) [inline mayastor/tests/aio_unmap.rs:64]: Could you please use a constant to define the amount of volumes to be created/tested ?

**Comment by mtzaurus** [comment]
I have a generic question: is it worth adding a testcase to read unmapped range and make sure we read zeroes from unmapped range ?

**Comment by cjones1024** [comment]
> 
> 
> I have a generic question: is it worth adding a testcase to read unmapped range and make sure we read zeroes from unmapped range ?

Good Idea, Will add

File: mayastor/tests/aio_unmap.rs (hunk 4 above)
Comment (cjones1024) [reply mayastor/tests/aio_unmap.rs]: Done

File: mayastor/tests/aio_unmap.rs (hunk 3 above)
Comment (cjones1024) [reply mayastor/tests/aio_unmap.rs]: Done

**Review by tiagolobocastro — APPROVED** [review]

**Comment by cjones1024** [comment]
bors try

**Comment by cjones1024** [comment]
bors try

File: mayastor/tests/aio_unmap.rs [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+        .await
+        .expect("failed awaiting write_zeroes completion")
+    {
+        return Ok(());
+    }
+
+    Err(Error::new(ErrorKind::Other, "write_zeroes failed"))
+}
+
+fn setup() -> &'static MayastorTest<'static> {
```
Comment (gila) [inline mayastor/tests/aio_unmap.rs:146]: Can we use Compose:::MaystorTest for this?

**Comment by cjones1024** [comment]
bors check

**Comment by cjones1024** [comment]
bors try

**Comment by jonathan-teh** [comment]
This only tests the "file" codepath in the aio bdev module, are there plans to test the "block" codepath too? Perhaps not here as it's tricky to set up a temporary block device and clean up afterwards.

**Review by jonathan-teh — APPROVED** [review]

**Comment by cjones1024** [comment]
bors merge

**Pull request merged**